### PR TITLE
TransformControls v2

### PR DESCRIPTION
**Pull request by arodic**
Here is a long awaited and relatively big update to TransformControls!

This PR fixes issues reported in #13806 #5489 #5379 #10916 #9312 #7947

This is pretty much a full rewrite on top of the original API. It includes a bunch of small improvements, simplifications and a few new features.

Some of the updates are:

- There is no need to use setter functions since changing relevant properties will automatically triger change event. For example. `control.setMode('scale')` should now be `control.mode = 'scale'`. Moreover, calling `control.update()` is no longer needed since updates are done automatically with `updateMatrixWorld`.

- Transform gizmo now includes "helpers" that can serve as visual aid showing current transformation axis for example.

- Handling of axis orientation is much improved. Axes pointing towards camera will automatically be hidden while axes occluded by other axes will flip to front side of the gizmo.

transform-axis-hide

- Rotation at grazing angles is now possible. Also, dragging the object center (between axes) will rotate the object around arbitrary axis perpendicular to drag direction.

transform-rotate

- Scale mode now correctly scales the object at rate relative to drag gesture in the viewport. Also, additional handles for XY, YZ and XZ scale are added.

transform-scale

- Transform gizmo looks slightly better and flips/hides axes depending on angle.

transform-translate

### CODE REVIEW

**Comment by mrdoob** [comment]
Live link: https://rawgit.com/arodic/three.js/transform-controls/editor/

**Comment by arodic** [comment]
I will! We should also consider adding deprication warnings to old Geometry constructors.

File: editor/js/Viewport.js [hunk 1]
```diff
@@ -285,19 +284,19 @@ var Viewport = function ( editor ) {
 
 	signals.transformModeChanged.add( function ( mode ) {
 
-		transformControls.setMode( mode );
```
Comment (mrdoob) [inline editor/js/Viewport.js]: I think I prefer keeping these methods, it's easier to keep backwards compatibility when changing the API...

Comment (arodic) [reply editor/js/Viewport.js]: I was hoping to maintain TransformControls as something integral to an editor built with my custom UI framework.

This framework relies heavily on a data binding implementation which avoids setter functions by design.

I have nothing against having specialized setter functions as long as the object.defined property getters/setters work as designed here. I can just remove deprication warnings and leave the functions hanging there... Sounds good?

Comment (arodic) [reply editor/js/Viewport.js]: @mrdoob I don't know if you've noticed by the way. The functions are still there but throw a deprication warning.

**Comment by arodic** [comment]
@mrdoob I removed deprecation warnings for setter functions and reverted the examples to use them.

**Comment by haroldiedema** [comment]
I'm using your TransformControls in the level editor of my game engine but had to modify some things to get it working properly. The reason for this is because I'm using WebWorkers pretty excessively and there is no `document`, let alone mouse events directly available in a worker context.



So I'm wondering if it would be possible for you to separate the logic that handles mouse events from the  event listeners. Basically what I'm asking is that you change things like:
```js
function onPointerHover(event) {
    if ( scope.object === undefined || scope.dragging === true || ( event.button !== undefined && event.button !== 0 ) ) return;
```
to simply:
```js
function onPointerHover(mouseX, mouseY) {
    if ( scope.object === undefined || scope.dragging === true ) return;
    // ... do stuff here.
```
And do the logic to check which button is pressed etc, somewhere in the same segment where you also bind the event listeners to the mouse events. This way, if people use THREE in a worker context like myself, it takes a lot less effort hacking in the code to get this working properly.

If you find this too much hassle, I'm happy to file a PR once this one gets merged.

**Comment by arodic** [comment]
Interesting proposal @haroldiedema

I think all controls should handle listeners separately from control logic. Not necessarily because worker context but because the listener logic is significant and complicated enough to have it's own abstraction. 

I noticed that each control has its own implementation identical in intention but slightly varied in implementation. Also, various UX bugs emerged multiple times for nearly identical reasons in different controls.

Having a single implementation for listeners that normalizes mouse/touch and serves the purpose for all controls would greatly improve maintenability of controls and open a possibility for more sophisticated scene manipulation tools.

Anyway...

The thing you proposed should be easy to implement. Would it be OK, if handler arguments sit in an object similar to native event object? I know you can post JSON to a worker but not sure if doing so at real-time rate is an issue.

**Comment by arodic** [comment]
Hey @haroldiedema I separated the event handlers from control logic but I imagine you will still need to make some changes to make it work inside worker context.

**Comment by haroldiedema** [comment]
Looks good! 👍 

If it were completely up to me, I would instead create a separate "controller" that you would optionally "inject" into the constructor of the entire class though. One works for the normal 'document' or 'domElement' context, the other works for custom events.

(overly) simplified usage would be something like this:
```js
// Normal use:
let transformControls = new TransformControls(camera, myCanvas);

// Worker use:
let transformControls = new TransformControls(camera, myController);

// Where myController is something like this:
class MyController extends EventEmitter
{
    constructor(myMouseDriver)
    {
        myMouseDriver.on('right-down', () => this.emit('mousedown', {button: 2});
        myMouseDriver.on('right-up', () => this.emit('mouseup', {button: 2});
        myMouseDriver.on('move', (x, y) => this.emit('mousemove', {x: x, y: y});
        // etc... (you probably want real function references so you can remove listeners again, but thats up to the developer implementing his/her own version of this controller.
    }
}
```

Meanwhile inside your code, there is something similar to this but one that works for DOM event listeners like you do it now with the part that describes the following:
```js
domElement.addEventListener( "mousedown", onPointerDown, false );
domElement.addEventListener( "touchstart", onPointerDown, false );
domElement.addEventListener( "mousemove", onPointerHover, false );
domElement.addEventListener( "touchmove", onPointerHover, false );
// etc..
```

So that would look like something like this:
```js
class DefaultInputController extends EventEmitter
{
    constructor(domElement)
    {
        let domElement = domElement || document;

        domElement.addEventListener( "mousedown", this.handleMouseDown);
    }

    handleMouseDown(event)
    {
        this.emit('mousedown', {button: 2}); // Note how this signature is the exact same as the one in "MyController".
    }
}
```

So the constructor of `TransformControls` would go from this:
```js
THREE.TransformControls = function ( camera, domElement ) {

        THREE.Object3D.call( this );

        domElement = ( domElement !== undefined ) ? domElement : document;
        // ...
}
```

to this:
```
THREE.TransformControls = function ( camera, domElementOrController ) {

        THREE.Object3D.call( this );
        
        if (domElementOrController === undefined) {
            inputController = new DefaultInputController(document);
        } else if (domElementOrController instanceof HTMLElement) {
            inputController = new DefaultInputController(domElementOrController);
        } else {
            inputController = domElementOrController;
        }

        // ...
        // ...

        inputController.on('mousemove', this.pointerDown);

        // ...

        this.pointerDown = function (pointer) {
            let x = pointer.x,
                 y = pointer.y; 

        }
}

// ...


```

**Comment by arodic** [comment]
@haroldiedema those ES6 classes look really nice! Its the only way forward. I really hope #9562 goes through soon.

**Comment by pailhead** [comment]
This is a super sweet example, i hope it gets merged soon!

**Comment by brettkromkamp** [comment]
Real improvement to the current TransformControls functionality so looking forward to this being merged. Pity it was bumped from r94 into r95.

**Comment by arodic** [comment]
@brettkromkamp I think it is OK to wait this one out a little since it is a full rewrite. In the meantime, if people want to test the branch and make sure controls work OK, it would be appreciated. Please report bugs here if you find any!

Also, looking at this after some time has passed, I'm wondering if two rotation modes are a the most elegant solution? Maybe it is better to go with 3dsMax style (linear gesture only) which works at any camera angle.

**Comment by easyfrog** [comment]
It's a pretty nice solution. and is it support multiple selection now ?

**Comment by NicolasRannou** [comment]
Looking forward to it! I will use it in my project, after adapting to typescript!

**Comment by NicolasRannou** [comment]
Any ETA on this one?

**Comment by arodic** [comment]
¯\_(ツ)_/¯

If there is anything I can do to expedite, I'll be available in about a week.

Perhaps you could just pull this into your branch if you need it right away.

**Comment by pailhead** [comment]
>Perhaps you could just pull this into your branch if you need it right away. 

It is, after all, an example :) Where the file(s) come from does not matter. @arodic would you consider pushing this to it's own repo?

**Comment by arodic** [comment]
@pailhead that's the idea... I'll get to it soon. Maybe.

**Comment by pailhead** [comment]
I think it would be much easier to clone or install if it lived in its own repo.

**Comment by trusktr** [comment]
A `mouseOver` event would be sweet. For example, in my app, I want to have logic that prevents the controls from auto-hiding after a timeout if you've hovered on it.

**Comment by arodic** [comment]
@trusktr TransformControls does not mess with external classes. My guess it has something to do with listeners on domElement. But to my knowledge, TransformControls v2 has the same listeners and reverse compatible API. I can look into this but it would be helpful if you could do some debugging on your end or provide working example.

**Comment by mrdoob** [comment]
Sorry for the delay 😇

**Comment by mrdoob** [comment]
Many thanks!

**Comment by arodic** [comment]
@trusktr maybe [this](https://github.com/arodic/three.js/commit/fb867488d87574a85d6236ec59b3009ca6a1c778) solve your issue.

**Comment by AustinDuvin** [comment]
I am having the same problem and your proposed fix did nothing. Is the way we instantiate a TransformControl different now?

**Comment by arodic** [comment]
@AustinDuvin see discussion #14619

Are you using OrbitControls or some other?

If you want to make TransformControls work nicely with any other control, look at editor Viewport.js and how it handles multiple controls.

The reason TransformControls used to work alongside OrbitControls with no setup was quite an ugly hack that involved event.stopPropagation(). Instead of re-introducing the hack, I provided a simple way to rig multiple controls in #14619 (see TransformControls example in that branch)

If you provide me with a code sample of your controls I can assist you further.

**Comment by AustinDuvin** [comment]
Okay, I implemented the changes and now it's working fine. It seems to have had some problems depending on when the controls were created and added to the scene. It may just be that the way it interacts with other parts of the software I'm working on caused some issues where it tried to create the controls before the scene had fully loaded or been created. I'm not entirely sure what the cause of the problem was, but it is no longer an issue. Thank you.

**Pull request merged**